feat: Making Quartz available offline by making it a PWA

adambrangenberg opened the pull request:
Initially I implemented it using the normal API but ran into some weird edge cases. Later, I've used workbox-cli to create the service worker because according to Googles Documentation the regular API is very bare-bones and has a few edge cases which get fixed by using Workbox. This has the advantage that we won't need to test other edge cases for different setups, which may not be reproducible for us in the future.

The manifest.json files are getting generated in the build process. I'm writing it to public with fs.writeSync. If there's a quartz-specific way to do write files to public, I'd be great to use that. One thing which we could consider is adding a separate colour for the theme of the browser. Currently, it uses the background colour of the lightmode for that. But many browsers also ignore that colours, so it may be unnecessary complexity.

I've documented the `workbox.config.ts`, the new `description` field for the manifest and the umami analytics field since it wasn't documented yet.

Edit: typos

adambrangenberg committed:
Adding PWA and chaching for offline aviability

adambrangenberg committed:
renamed workbox config to fit Quartz' scheme

adambrangenberg committed:
Documenting new configuration

adambrangenberg committed:
Added missig umami documentation

adambrangenberg committed:
Fixed formatting so the build passes, thank you prettier :)

adambrangenberg commented:
Oh, and I also added a second Quartz Icon as an SVG, so it can be used for any icon size by the browser :)

jackyzha0 reviewed (commented):
I'm a bit more hesitant to ship this to main now seeing as the install size for `workbox-cli` is ~50MB. And the bundle size of `sw.js` is pretty big too and I prefer wouldn't ship by default unless the user opts-in for it. I wonder if there's a way we can just write the docs for how to enable PWA for your Quartz but don't ship it.

jackyzha0 commented:
(side note: happy to merge docs updates in another PR)

adambrangenberg commented:
> I'm a bit more hesitant to ship this to main now seeing as the install size for `workbox-cli` is ~50MB. And the bundle size of `sw.js` is pretty big too and I prefer wouldn't ship by default unless the user opts-in for it. I wonder if there's a way we can just write the docs for how to enable PWA for your Quartz but don't ship it.

The easiest thing would probably be to just don't run it in the build process?

adambrangenberg commented:
Or we could use the bare-bones APIs, but well edge cases

jackyzha0 commented:
Can you link to the edge-cases? I think Quartz is simple enough that it probably _shouldn't_ matter but let's make sure. I would prefer to go with the bare-bones API

adambrangenberg commented:
It was some specific site I can't find anymore which says there are edge cases with the API in general. Will go through my browsing history later in case it's really important.

In the "main tutorial" on web.dev it's mentioned how Safari for example does a lot of things differently or doesn't support them in the first place: https://web.dev/learn/pwa/

adambrangenberg committed:
specified caching strategies to improve performance

adambrangenberg committed:
formatting...

adambrangenberg commented:
I've specified the chaching strategies explicitly to get some performance benefits. This would also make Navigation on non-SPA sites smoother since only the actual content is getting requested via the Network and the rest fetched from cache. 

In case you didn't read the PWA overview yet, here's what they mean:

- NetworkFirst (used for documents): Gets it from the server, if not available from cache
- CacheFirst (used for images and scripts): Gets is from the cache, if not available from the Network
- StaleWhileRevalidate (used for other stuff): Gets if from the cache and in the background from the Network. Next time the page loads the new content gotten from the Network is used and so on

Since Scripts Cache only has a specified lifetime of one day, it may make sense to switch it to StaleWhileRevalidate per default and increase the lifetime so that people can use it offline for a longer time.

adambrangenberg committed:
fixing "404 manifest.json not found" on subdirectories by adding a / to manifestpath

geekodour commented:
Thanks for the work on this @adambrangenberg! (ty from discord here)

Just some thoughts, not suggestions:

I've not tested this yet, but how does PWAs handle updates? Eg. I pushed a new page say `some_page.md`, will the PWA automatically pull it when I load it while online? Or in this case user would need to explicitly visit the page `some_page.html` ?

>  I wonder if there's a way we can just write the docs for how to enable PWA for your Quartz but don't ship it.

I think we can put the `manifest` creation and workbox generation in `quartz/build.ts` inside some flag/conditional and then have `workbox.config.ts` checked in. Changes in `quartz/components/Head.tsx`, `quartz/plugins/emitters/componentResources.ts` can be manually documented along with installation of `workbox-cli`

Addition of the `description` field can be skipped info, just the `pageTitle` reused for `description` in the `manifest` creation should could do.

adambrangenberg commented:
As long you run the build script when adding it, yup it should work. Same for updated content which is why I made content NetworkFirst by default. 

The only weird thing which I found yesterday is that when rebuilding Quartz (like when you edit an page and it automatically rebuilds) doesn't run the sw building thing too and deletes it from the public directory.

adambrangenberg committed:
turning it into a plugin

adambrangenberg committed:
Removed Workbox-cli and updated @types/node

adambrangenberg committed:
Added Serviceworkercode to offline.ts

adambrangenberg committed:
formatting

adambrangenberg committed:
Merge pull request #1 from jackyzha0/v4

merge new changed

adambrangenberg commented:
@jackyzha0 I changed it now so that we only need workbox-sw which imports workbox-recipies. It now uses the default strategies for the specific resource type. 

Based on the chat on Discord, I've removed precaching so that pages only get cached/updated when you visit them.

adambrangenberg commented:
I've also packaged it as a emit plugin now

adambrangenberg committed:
Removing workbox from docs

adambrangenberg commented:
Also should I document the plugin? I've seen that a few things are undocumented and Offline doesn't really have any configuration to it

jackyzha0 commented on docs/configuration.md:
````diff
@@ -78,5 +80,3 @@ transformers: [
   Plugin.Latex({ renderEngine: "katex" }), // specify some options
 ]
 ```
-
-If you'd like to make your own plugins, read the guide on [[making plugins]] for more information.
````
why did this get removed?

adambrangenberg replied:
I had some issues with the current Quart branch and the branch before. Git probably removed it while resolving it?

jackyzha0 commented on quartz/build.ts:
```diff
@@ -17,6 +17,9 @@ import { glob, toPosixPath } from "./util/glob"
 import { trace } from "./util/trace"
 import { options } from "./util/sourcemap"
 import { Mutex } from "async-mutex"
+import { execSync } from "child_process"
+import { writeFileSync } from "fs"
+import { copyFileSync } from "node:fs"
```
```suggestion
```

unused deps?

jackyzha0 commented on quartz/plugins/emitters/offline.ts:
```diff
@@ -0,0 +1,54 @@
+
+export const Offline: QuartzEmitterPlugin = () => {
+  return {
+    name: "Offline",
```
```suggestion
    name: "OfflineSupport",
```

jackyzha0 commented on quartz/plugins/emitters/offline.ts:
```diff
@@ -0,0 +1,54 @@
+        start_url: "/",
+      }
+
+      const serviceWorker =
+        "importScripts('https://storage.googleapis.com/workbox-cdn/releases/6.4.1/workbox-sw.js');" +
+        "const {pageCache, imageCache, staticResourceCache, googleFontsCache, offlineFallback} = workbox.recipes;" +
+        "pageCache(); googleFontsCache(); staticResourceCache(); imageCache();offlineFallback();"
```
can we use string interpolation here?

adambrangenberg replied:
Sure, it'll just be more bytes,

adambrangenberg replied:
@jackyzha0 yes or no now?

jackyzha0 replied:
i prefer string interpolation for readability reasons

jackyzha0 commented on quartz/plugins/emitters/offline.ts:
```diff
@@ -0,0 +1,54 @@
+        "const {pageCache, imageCache, staticResourceCache, googleFontsCache, offlineFallback} = workbox.recipes;" +
+        "pageCache(); googleFontsCache(); staticResourceCache(); imageCache();offlineFallback();"
+
+      return [
+        await emit({
+          content: JSON.stringify(manifest),
+          slug: path.join("manifest") as FullSlug,
+          ext: ".json",
+        }),
+        await emit({
+          content: serviceWorker,
+          slug: path.join("sw") as FullSlug,
+          ext: ".js",
+        }),
+      ]
```
```suggestion
      return await Promise.all([
        emit({
          content: JSON.stringify(manifest),
          slug: path.join("manifest") as FullSlug,
          ext: ".json",
        }),
        emit({
          content: serviceWorker,
          slug: path.join("sw") as FullSlug,
          ext: ".js",
        }),
      ])
```

jackyzha0 commented on workbox.config.ts:
```diff
@@ -0,0 +1,77 @@
+      },
+    },
+    {
+      // Scrips
```
```suggestion
      // Scripts
```

jackyzha0 commented on workbox.config.ts:
```diff
@@ -0,0 +1,77 @@
+  // Don't change that unless you know what you're doing
+  globDirectory: "public/",
+  swDest: "public/sw.js",
+  ignoreURLParametersMatching: [/^utm_/, /^fbclid$/],
```
is there a way to ignore all params? iirc Quartz doesn't have any features that use query parameters for routing

adambrangenberg replied:
I removed the workbox config, not sure why it's still displayed here?

jackyzha0 replied:
try a `git rm workbox.config.ts`

adambrangenberg replied:
I opened a codespace and removed it there. Should be gone now :)

adambrangenberg committed:
applied suggestions

jackyzha0 commented on quartz/components/Head.tsx:
```diff
@@ -25,7 +25,9 @@ export default (() => {
         <meta property="og:height" content="675" />
+        <meta name="theme-color" content="#faf8f8" />
         <link rel="icon" href={iconPath} />
+        <link rel="manifest" href="/manifest.json" />
```
use `baseDir` here, accessing something via absolute path to root is incorrect for any Quartz hosted on a subpath (e.g. `username.github.io/quartz`)

adambrangenberg replied:
What about the start_url in the Manifest then? I thought both of them behave based on the domain?

jackyzha0 commented on quartz/plugins/emitters/offline.ts:
```diff
@@ -0,0 +1,53 @@
+      return Promise.all([
+        emit({
+          content: JSON.stringify(manifest),
+          slug: path.join("manifest") as FullSlug,
```
```suggestion
          slug: "manifest" as FullSlug,
```

jackyzha0 commented on quartz/plugins/emitters/offline.ts:
```diff
@@ -0,0 +1,53 @@
+        }),
+        emit({
+          content: serviceWorker,
+          slug: path.join("sw") as FullSlug,
```
```suggestion
          slug: "sw" as FullSlug,
```

adambrangenberg committed:
Removed path.join for sw path

Co-authored-by: Jacky Zhao <j.zhao2k19@gmail.com>

adambrangenberg committed:
Removed path.join for manifest path

Co-authored-by: Jacky Zhao <j.zhao2k19@gmail.com>

adambrangenberg committed:
Removing path module import

adambrangenberg committed:
Added absolute path to manifests start_url and manifest "import" using baseUrl

jackyzha0 commented on quartz/components/Head.tsx:
```diff
@@ -25,7 +25,9 @@ export default (() => {
         <meta property="og:height" content="675" />
+        <meta name="theme-color" content="#faf8f8" />
         <link rel="icon" href={iconPath} />
+        <link rel="manifest" href="${cfg.baseUrl}/manifest.json" />
```
```suggestion
        <link rel="manifest" href="https://${cfg.baseUrl}/manifest.json" />
```

jackyzha0 replied:
`baseUrl` doesn't have the protocol

jackyzha0 reviewed (approved):
looks good, just a few more small nits. could you also add a docs page for this? ready to merge otherwise

adambrangenberg committed:
Adding protocol to baseurl

Co-authored-by: Jacky Zhao <j.zhao2k19@gmail.com>

adambrangenberg commented:
If you tell me where in the Docs to link it, sure! :D

jackyzha0 commented:
a new file under `/docs/features` would be fine with the `#plugin/emitter` tag. You can link to it from the main feature listing on the home index page

adambrangenberg committed:
Adding protocol to start_url too then

adambrangenberg committed:
formatting...

adambrangenberg committed:
Adding fallback page

adambrangenberg committed:
Documenting offline plugin

adambrangenberg commented:
Done, also added the offline fallback page :)

adambrangenberg committed:
formatting...

jackyzha0 commented on docs/offline.md:
```diff
@@ -0,0 +1 @@
+The requested page wasn't cache already.
```
```suggestion
The requested page wasn't cached.
```

this shouldn't be a markdown file, should be emitted like the 404 page is. i also think it's ok to just not show anything for non-cached pages

jackyzha0 commented on docs/features/offline access.md:
```diff
@@ -0,0 +1,32 @@
+---
```
```suggestion
---
title: "Offline Access (PWA)"
```

jackyzha0 commented on docs/features/offline access.md:
```diff
@@ -0,0 +1,32 @@
+  - plugin/emitter
+---
+
+This allows your Website to be accesible offline and be installed as an app. You can use it by adding `Plugin.Offline(),` to the `emitters` in `quartz.config.ts`
```
```suggestion
This plugin allows your website to be accessible offline and be installed as an app. You can use it by adding `Plugin.Offline(),` to the `emitters` in `quartz.config.ts`
```

jackyzha0 commented on docs/features/offline access.md:
```diff
@@ -0,0 +1,32 @@
+
+Whenever you visit a page it gets cached for offline use. Depending on the kind of content, the process for caching is diffent:
+
+- **Pages** (HTML, your converted Markdown files): Quartz first tries to get them over the Network. If that fails is goes back to the Cache
```
```suggestion
- **Pages** (HTML, your converted Markdown files): Quartz first tries to get them over the Network. If that fails, your browser attempts to fetch it from the cache.
```

jackyzha0 commented on docs/features/offline access.md:
```diff
@@ -0,0 +1,32 @@
+Whenever you visit a page it gets cached for offline use. Depending on the kind of content, the process for caching is diffent:
+
+- **Pages** (HTML, your converted Markdown files): Quartz first tries to get them over the Network. If that fails is goes back to the Cache
+- **Static Resources** (Fonts, CSS Styling, JavaScript): Quartz gets them from cache and updates the cache over the Network in the background. Next time the page loads this new cache is being used
```
```suggestion
- **Static Resources** (Fonts, CSS Styling, JavaScript): Quartz uses cached resources by default and updates the cache over the network in the background.
```

jackyzha0 commented on docs/features/offline access.md:
```diff
@@ -0,0 +1,32 @@
+
+- **Pages** (HTML, your converted Markdown files): Quartz first tries to get them over the Network. If that fails is goes back to the Cache
+- **Static Resources** (Fonts, CSS Styling, JavaScript): Quartz gets them from cache and updates the cache over the Network in the background. Next time the page loads this new cache is being used
+- **Images**: Those are getting saved once and then served from Cache. The amount of images is limited to 60 Images and are valid for 30 Days
```
```suggestion
- **Images**: Images are saved once and then served from cache. Quartz uses a limited cache of 60 images and images remain in the cache for 30 days
```

jackyzha0 commented on docs/features/offline access.md:
```diff
@@ -0,0 +1,32 @@
+
+You can edit the fallback page by changing the `offline.md` file in the root of your `content` directory
+
+## App
```
```suggestion
## Progressive Web App (PWA)
```

jackyzha0 commented on docs/features/offline access.md:
```diff
@@ -0,0 +1,32 @@
+
+Progressive Web Apps can have [many properties](https://developer.mozilla.org/en-US/docs/Web/Manifest). We're only going to mention the ones Quartz supports by default, however you can edit the offline plugins file to add more in case required.
+
+### Properties taken from configuration
```
```suggestion
```

jackyzha0 commented on docs/features/offline access.md:
```diff
@@ -0,0 +1,32 @@
+### Properties taken from configuration
+
+- **icons**: the `icon.svg` file in the `quartz/static` directory is used for all the icons. This makes it easier to scale the image since you don't need to provide an png for every size
+- **name** and **short_name**: For this we use the `pageTitle` you configured in `quartz.config.ts`
```
```suggestion
- **name**, **short_name**: Uses the `pageTitle` configured in `quartz.config.ts`
```

jackyzha0 commented on docs/features/offline access.md:
```diff
@@ -0,0 +1,32 @@
+
+- **icons**: the `icon.svg` file in the `quartz/static` directory is used for all the icons. This makes it easier to scale the image since you don't need to provide an png for every size
+- **name** and **short_name**: For this we use the `pageTitle` you configured in `quartz.config.ts`
+- **description**: For this we use the `description` you configured in `quartz.config.ts`
```
```suggestion
- **description**: Uses the `description` configured in `quartz.config.ts`
```

jackyzha0 commented on docs/features/offline access.md:
```diff
@@ -0,0 +1,32 @@
+- **icons**: the `icon.svg` file in the `quartz/static` directory is used for all the icons. This makes it easier to scale the image since you don't need to provide an png for every size
+- **name** and **short_name**: For this we use the `pageTitle` you configured in `quartz.config.ts`
+- **description**: For this we use the `description` you configured in `quartz.config.ts`
+- **background_color** and **theme_color**: For this we use the `lightMode.light` color you configured in `quartz.config.ts`. However many browsers don't use that functionality
```
```suggestion
- **background_color**, **theme_color**: Uses the `lightMode.light` color configured in `quartz.config.ts`.
```

jackyzha0 commented on docs/features/offline access.md:
```diff
@@ -0,0 +1,32 @@
+- **name** and **short_name**: For this we use the `pageTitle` you configured in `quartz.config.ts`
+- **description**: For this we use the `description` you configured in `quartz.config.ts`
+- **background_color** and **theme_color**: For this we use the `lightMode.light` color you configured in `quartz.config.ts`. However many browsers don't use that functionality
+- **start_url**: For this we use the `baseUrl` you configured in `quartz.config.ts`
```
```suggestion
- **start_url**: Uses the `baseUrl` configured in `quartz.config.ts`
```

jackyzha0 reviewed (commented):
grammar fixes, note about offline.md

adambrangenberg committed:
merge suggestion

Co-authored-by: Jacky Zhao <j.zhao2k19@gmail.com>

adambrangenberg committed:
merge suggestion

Co-authored-by: Jacky Zhao <j.zhao2k19@gmail.com>

adambrangenberg committed:
merge suggestion

Co-authored-by: Jacky Zhao <j.zhao2k19@gmail.com>

adambrangenberg committed:
merge suggestion

Co-authored-by: Jacky Zhao <j.zhao2k19@gmail.com>

adambrangenberg committed:
merge suggestion

Co-authored-by: Jacky Zhao <j.zhao2k19@gmail.com>

adambrangenberg committed:
merge suggestion

Co-authored-by: Jacky Zhao <j.zhao2k19@gmail.com>

adambrangenberg committed:
merge suggestion

Co-authored-by: Jacky Zhao <j.zhao2k19@gmail.com>

adambrangenberg committed:
merge suggestion

Co-authored-by: Jacky Zhao <j.zhao2k19@gmail.com>

adambrangenberg committed:
merge suggestion

Co-authored-by: Jacky Zhao <j.zhao2k19@gmail.com>

adambrangenberg committed:
merge suggestion

Co-authored-by: Jacky Zhao <j.zhao2k19@gmail.com>

adambrangenberg committed:
merge suggestion

Co-authored-by: Jacky Zhao <j.zhao2k19@gmail.com>

adambrangenberg committed:
merge suggestion

Co-authored-by: Jacky Zhao <j.zhao2k19@gmail.com>

adambrangenberg committed:
formatting...

jackyzha0 commented:
not sure if you saw my earlier comment about the `offline.md` file: 

> this shouldn't be a markdown file, should be emitted like the 404 page is. i also think it's ok to just not show anything for non-cached pages which might be the easiest option here

adambrangenberg commented:
I didn't see it, thank you. Could we just use the 404 file too?

jackyzha0 commented:
i don't think 404 is accurate, it's just not cached. we could just let not-cached behaviour fall through to default browser behaviour

adambrangenberg commented:
I'm not sure what you mean, but will just add a offline page then

jackyzha0 commented:
What I'm trying to say is that we should just remove `offlineFallback` because I think showing the default 'This page can't be reached' is more indicative of what's actually going on then serving an offline page

adambrangenberg committed:
Fixing manifest path, all these nits hiding the actual issues .-.

adambrangenberg committed:
Offline fallback page through plugins, most things taken from 404 Plugin

adambrangenberg committed:
adding Offline Plugin to config

adambrangenberg committed:
formatting...

adambrangenberg commented:
Oh okay

jackyzha0 commented:
Can you remove `docs/offline.md` and remove `Offline` as a default option? I'd like it to be opt-in if possible. Ready to merge when done

adambrangenberg commented:
Sure, we had a school event today. Will make it once I'm on my Laptop

adambrangenberg commented:
Should the fallback page stay though?

jackyzha0 commented:
I'll leave it up to you, I think it's better without the fallback page but I personally don't use PWAs often enough to know what best practice is

adambrangenberg commented:
It's closer to an actual App, so depends on the goal

ref: https://web.dev/offline-fallback-page/

adambrangenberg committed:
Turned offline off as default and removed offline.md

adambrangenberg commented:
> Can you remove `docs/offline.md` and remove `Offline` as a default option? I'd like it to be opt-in if possible. Ready to merge when done

Done :)

raeglan commented:
If I understood correctly this feature was reverted, right? 
What was there reason for it? Is there something that I or someone else can do to to make it available again?

adambrangenberg commented:
> If I understood correctly this feature was reverted, right? What was there reason for it? Is there something that I or someone else can do to to make it available again?

After nitpicking for ages, Jacky just changed his mind about adding something (I think it was the description or baseUrl? Not sure, the conversation is in the Discord) in the Header last second. Without the thing, the whole feature is not possible to implement because it wouldn't find the Manifest or something
